sys/riscv/riscv/trap.c

```diff
@@ -319,7 +317,7 @@ do_trap_supervisor(struct trapframe *frame)
 	exception = frame->tf_scause & SCAUSE_CODE;
 	if ((frame->tf_scause & SCAUSE_INTR) != 0) {
 		/* Interrupt */
-		intr_irq_handler(frame);
+		intr_irq_handler(frame, INTR_ROOT_IRQ);
```

Comment:
RISC-V is not calling a wrapper function, this is the exact same `intr_irq_handler` that ARM{,64} call and equivalent code is generated on these platforms here. The only difference is that one source file is asm while the other is C.

Reply:
I've always assumed it's because Arm's VBAR does the interrupt-vs-exception dispatch in hardware thanks to having multiple trap vectors, whereas RISC-V has just the one in direct mode which requires demuxing in software, and it's best to keep the assembly minimal, so it's done in C.